From 30866d05664ce01f0818d24475644b33b140a3db Mon Sep 17 00:00:00 2001
From: Jeff Vander Stoep <jeffv@google.com>
Date: Tue, 18 Aug 2020 14:00:05 +0200
Subject: [PATCH] cargo: Don't list-files in bare git repo

This change was mistakenly removed in aosp/1403447.

Bug: 165026104
Test: TH along with 1.45.2 upgrade
Change-Id: I8fe7223dbfa085794dfeea45692ea684ea0f24d8
---
 src/tools/cargo/src/cargo/sources/path.rs | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/src/tools/cargo/src/cargo/sources/path.rs b/src/tools/cargo/src/cargo/sources/path.rs
index cf406e8ddb..0f75fef750 100644
--- a/src/tools/cargo/src/cargo/sources/path.rs
+++ b/src/tools/cargo/src/cargo/sources/path.rs
@@ -191,6 +191,15 @@ impl<'cfg> PathSource<'cfg> {
         let index = repo
             .index()
             .chain_err(|| format!("failed to open git index at {}", repo.path().display()))?;
+
+        // If rustc is checked into a bare git repo, we will error out below.
+        // The fallback path still works, so pretend it didn't find a git repo rather than
+        // erroring out.
+        // See rust-lang/cargo#7876.
+        if repo.is_bare() {
+            return Ok(None)
+        }
+
         let repo_root = repo.workdir().ok_or_else(|| {
             anyhow::format_err!(
                 "did not expect repo at {} to be bare",
-- 
2.28.0.220.ged08abb693-goog

